crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,253 @@
+use anyhow;
+use async_dropper::AsyncDrop;
+use tokio::task::JoinHandle as tokioJoinHandle;
+
+use async_trait::async_trait;
+use futures_util::{
+    task::{Context, Poll},
+    Future,
+};
+
+use std::pin::Pin;
+use tokio::signal::unix::{Signal, SignalKind};
+use tokio::sync::{mpsc, OnceCell};
+
+#[async_trait]
+pub trait Task {
+    /// A set of tasks which are all executed on the same 'execution backend'
+    type TaskSet;
+
+    /// Performs the unit of work for the task
+    async fn run_task(
+        self,
+        signals: mpsc::Receiver<Signal>,
+        task_group: &Self::TaskSet,
+    ) -> anyhow::Result<()>;
+
+    /// Spawns the task
+    async fn spawn_task(self) -> JoinHandle;
+}
+
+pub struct JoinHandle {
+    task: OnceCell<tokioJoinHandle<()>>,
+    allowed_signal: SignalKind,
```

Comment:
I think you've misunderstood the purpose of signals here. A signal is used to notify the task of some external action (commonly terminating). 

But instead of abruptly aborting the Tokio task (which doesn't leave any room for graceful termination / cleaning up of resources), its the implementator of the `Task` trait responsibility to ensure that when sent a termination signal, `run_task` returns. 

The intended use would be that the signal would be queued in the `mpsc::Sender`, with the implementator using the `mpsc::Recieve` to handle signals (including termination).

Reply:
Perhaps we should have an `abort` on a configurable timeout?